File: indicators.js
const EMA = require('technicalindicators').EMA


class Indicators {
    constructor() {
        // Wave Trend default parameters
        this.n1 = 10;  // Channel Length
        this.n2 = 21;  // Average Length
        this.obLevel1 = 60;   // Overbought Level 1
        this.obLevel2 = 53;   // Overbought Level 2
        this.osLevel1 = -60;  // Oversold Level 1
        this.osLevel2 = -53;  // Oversold Level 2
    }

    ema(values, period) {
        const k = 2 / (period + 1);
        let emaArray = [];
        emaArray[0] = values[0]; // Start EMA with first value
      
        for (let i = 1; i < values.length; i++) {
          emaArray[i] = values[i] * k + emaArray[i - 1] * (1 - k);
        }
        return emaArray;
    }

    // Calculate EMA (Exponential Moving Average)
    calculateEMA(source, length) {
        const alpha = 2 / (length + 1);
        
        let ema = source[0];
        const results = [ema];
        
        for (let i = 1; i < source.length; i++) {
            ema = alpha * source[i] + (1 - alpha) * ema;
            results.push(ema);
        }
        
        return results;
    }

    // Calculate SMA (Simple Moving Average)
    calculateSMA(data, period) {
        const results = [];
        
        for (let i = data.length - 1; i >= 0; i--) {
            if (i + period > data.length) continue;
            
            let sum = 0;
            for (let j = 0; j < period; j++) {
                sum += data[i + j];
            }
            results.push(sum / period);
        }
        
        return results;
    }

    pineEMA(source, length) {
        const alpha = 2 / (length + 1);
        let sum = source[0];  // Initialize with first value
        const result = new Array(source.length);
        result[0] = sum;
        
        for (let i = 1; i < source.length; i++) {
            sum = alpha * source[i] + (1 - alpha) * sum;
            result[i] = sum;
        }
        
        return result;
    }

    // Calculate Wave Trend
    calculateWaveTrend(candles) {
        if (candles.length < Math.max(this.n1, this.n2, 4)) {
            throw new Error(`Not enough candles to calculate Wave Trend. Need at least ${Math.max(this.n1, this.n2, 4)} candles, but got ${candles.length}`);
        }

        // Ensure candles are in newest-first order
        const orderedCandles = [...candles]; // Now newest-first
        //console.log('orderedCandles', orderedCandles[0]);
        // Step 1: Calculate AP (HLC3) with 8 decimal precision
        const ap = orderedCandles.map(candle => 
            Number(((candle.high + candle.low + candle.close) / 3))
        );

        // Step 2: Calculate ESA = EMA(AP, n1)
        const esa = this.pineEMA(ap, this.n1);

        // Step 3: Calculate D = EMA(abs(AP - ESA), n1)
        const absDiffs = ap.map((value, i) => 
            Number(Math.abs(value - esa[i]))
        );
        //console.log('absDiffs', absDiffs);
        const d = this.pineEMA(absDiffs, this.n1);

        // Step 4: Calculate CI = (AP - ESA) / (0.015 * D)
        const ci = ap.map((value, i) => {
            if (i < esa.length && i < d.length) {
                return Number(((value - esa[i]) / (0.015 * d[i])));
            }
            return 0;
        });

        // Step 5: Calculate TCI = EMA(CI, n2)
        const tci = this.pineEMA({
            period: this.n2,
            values: ci
        });

        // Step 6: Set WT1 = TCI
        const wt1 = tci;

        // Step 7: Calculate WT2 = SMA(WT1, 4)
        let wt2 = [];
        for (let i = 3; i < wt1.length; i++) {
            const sma = Number(((wt1[i] + wt1[i-1] + wt1[i-2] + wt1[i-3]) / 4).toFixed(8));
            wt2.push(sma);
        }

        // Get most recent values
        const currentValues = {
            ap: ap[ap.length - 1],
            esa: esa[esa.length - 1],
            d: d[d.length - 1],
            ci: ci[ci.length - 1],
            tci: tci[tci.length - 1],
            wt1: wt1[wt1.length - 1],
            wt2: wt2[wt2.length - 1]
        };

        // Debug logging for AP and EMA values
        console.log('\nLast 10 candles values:');
        for (let i = 0; i <= 9; i++) {
            const apValue = ap[i];
            const emaValue = esa[i];
            
            // Use the correct timestamp field
            const timestamp = candles[i].timestamp;
            const timeStr = new Date(timestamp).toLocaleTimeString('en-GB', {
                hour: '2-digit',
                minute: '2-digit',
                second: '2-digit'
            });
            
            console.log(`Time: ${timeStr} H: ${candles[i].high?.toFixed(6)} L: ${candles[i].low?.toFixed(6)} C: ${candles[i].close?.toFixed(6)} AP: ${apValue?.toFixed(6)} EMA: ${emaValue?.toFixed(6)} D: ${d[i]}`);
        }

        return {
            ...currentValues,
            overbought1: currentValues.wt1 >= this.obLevel1,
            overbought2: currentValues.wt1 >= this.obLevel2,
            oversold1: currentValues.wt1 <= this.osLevel1,
            oversold2: currentValues.wt1 <= this.osLevel2,
            crossOver: wt1.length > 1 && wt2.length > 1 && 
                (wt1[wt1.length-1] > wt2[wt2.length-1] && wt1[wt1.length-2] <= wt2[wt2.length-2]),
            crossUnder: wt1.length > 1 && wt2.length > 1 && 
                (wt1[wt1.length-1] < wt2[wt2.length-1] && wt1[wt1.length-2] >= wt2[wt2.length-2])
        };
    }

    // Method to calculate Wave Trend for a specific symbol and timeframe
    calculateWaveTrendForSymbol(candleStore, symbol, timeframe) {
        const candles = candleStore[symbol][timeframe];
        if (!candles || candles.length < Math.max(this.n1, this.n2)) {
            throw new Error(`Not enough candles for ${symbol} ${timeframe} to calculate Wave Trend`);
        }
        
        return this.calculateWaveTrend(candles);
    }

    // Method to set custom parameters
    setParameters({
        channelLength = 10,
        averageLength = 21,
        overbought1 = 60,
        overbought2 = 53,
        oversold1 = -60,
        oversold2 = -53
    } = {}) {
        this.n1 = channelLength;
        this.n2 = averageLength;
        this.obLevel1 = overbought1;
        this.obLevel2 = overbought2;
        this.osLevel1 = oversold1;
        this.osLevel2 = oversold2;
    }
}

module.exports = new Indicators();

File: indicators.py
import numpy as np
import pandas as pd
from ta.trend import EMAIndicator
from ta.momentum import RSIIndicator
from typing import List, Dict, Any
from datetime import datetime
import ta
import logging

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

class Indicators:
    def __init__(self):
        self.n1 = 10  # Channel Length
        self.n2 = 21  # Average Length
        self.ob_level1 = 60   # Overbought Level 1
        self.ob_level2 = 53   # Overbought Level 2
        self.os_level1 = -60  # Oversold Level 1
        self.os_level2 = -53  # Oversold Level 2
        self.sma50_period = 50
        self.sma200_period = 200

    def calculate_wave_trend(self, candles: List[Dict[str, Any]]) -> Dict[str, Any]:
        if len(candles) < max(self.n1, self.n2, 4):
            raise ValueError(f"Not enough candles to calculate Wave Trend. Need at least {max(self.n1, self.n2, 4)} candles, but got {len(candles)}")

        # Sort candles by timestamp in descending order (newest first)
        sorted_candles = sorted(candles, key=lambda x: x['timestamp'], reverse=True)
        
        # Convert sorted candles to pandas DataFrame and reverse order (oldest first)
        df = pd.DataFrame(sorted_candles[::-1])
        
        # Calculate AP (HLC3)
        df['ap'] = (df['high'] + df['low'] + df['close']) / 3
        
        # Calculate ESA = EMA(AP, n1)
        df['esa'] = df['ap'].ewm(span=self.n1, min_periods=self.n1, adjust=False).mean()
        
        # Calculate D = EMA(abs(AP - ESA), n1)
        df['d'] = abs(df['ap'] - df['esa'])
        df['d'] = df['d'].ewm(span=self.n1, min_periods=self.n1, adjust=False).mean()
        
        # Avoid division by zero
        df['d'] = df['d'].replace(0, 0.00001)
        
        # Calculate CI = (AP - ESA) / (0.015 * D)
        df['ci'] = (df['ap'] - df['esa']) / (0.015 * df['d'])
        
        # Calculate WT1 = EMA(CI, n2)
        df['wt1'] = df['ci'].ewm(span=self.n2, min_periods=self.n2, adjust=False).mean()
        
        # Calculate WT2 = SMA(WT1, 4)
        df['wt2'] = df['wt1'].rolling(window=4, min_periods=4).mean()
        
        # Reverse back to newest first
        df = df.iloc[::-1]

        # Debug logging
        #print('\nLast 10 candles values:')
        for i in range(min(10, len(df))):
            timestamp = datetime.fromtimestamp(df.iloc[i]['timestamp'] / 1000).strftime('%H:%M:%S')
            wt2_value = df.iloc[i]['wt2']
            wt2_str = f"{wt2_value:.6f}" if not pd.isna(wt2_value) else "N/A"
            
            print(f"Time: {timestamp} "
                  f"H: {df.iloc[i]['high']:.6f} "
                  f"L: {df.iloc[i]['low']:.6f} "
                  f"C: {df.iloc[i]['close']:.6f} "
                  f"AP: {df.iloc[i]['ap']:.6f} "
                  f"EMA: {df.iloc[i]['esa']:.6f} "
                  f"D: {df.iloc[i]['d']:.6f} "
                  f"WT1: {df.iloc[i]['wt1']:.6f} "
                  f"WT2: {wt2_str} "
                  f"OB1: {float(df.iloc[i]['wt1']) >= self.ob_level1} "
                  f"OB2: {float(df.iloc[i]['wt1']) >= self.ob_level2} "
                  f"OS1: {float(df.iloc[i]['wt1']) <= self.os_level1} "
                  f"OS2: {float(df.iloc[i]['wt1']) <= self.os_level2} "
                  f"CrossOver: {i > 0 and float(df.iloc[i]['wt1']) > float(df.iloc[i]['wt2']) and float(df.iloc[i-1]['wt1']) <= float(df.iloc[i-1]['wt2'])} "
                  f"CrossUnder: {i > 0 and float(df.iloc[i]['wt1']) < float(df.iloc[i]['wt2']) and float(df.iloc[i-1]['wt1']) >= float(df.iloc[i-1]['wt2'])}")

        # Get current values
        current = df.iloc[0]
        previous = df.iloc[1] if len(df) > 1 else None

        return {
            'ap': float(current['ap']),
            'esa': float(current['esa']),
            'd': float(current['d']),
            'ci': float(current['ci']),
            'wt1': float(current['wt1']),
            'wt2': float(current['wt2']) if not pd.isna(current['wt2']) else 0.0,
            'overbought1': float(current['wt1']) >= self.ob_level1,
            'overbought2': float(current['wt1']) >= self.ob_level2,
            'oversold1': float(current['wt1']) <= self.os_level1,
            'oversold2': float(current['wt1']) <= self.os_level2,
            'cross_over': previous is not None and float(current['wt1']) > float(current['wt2']) and float(previous['wt1']) <= float(previous['wt2']),
            'cross_under': previous is not None and float(current['wt1']) < float(current['wt2']) and float(previous['wt1']) >= float(previous['wt2'])
        }

    def calculate_moving_averages(self, candles: List[Dict[str, Any]]) -> Dict[str, Any]:
        if len(candles) < self.sma200_period:
            raise ValueError(f"Not enough candles to calculate SMAs. Need at least {self.sma200_period} candles, but got {len(candles)}")

        # Sort candles by timestamp in descending order (newest first)
        sorted_candles = sorted(candles, key=lambda x: x['timestamp'], reverse=True)
        
        # Convert sorted candles to pandas DataFrame
        df = pd.DataFrame(sorted_candles)
        
        # Calculate SMAs on the reversed data (oldest to newest)
        df['sma50'] = df['close'].iloc[::-1].rolling(window=self.sma50_period, min_periods=self.sma50_period).mean().iloc[::-1]
        df['sma200'] = df['close'].iloc[::-1].rolling(window=self.sma200_period, min_periods=self.sma200_period).mean().iloc[::-1]
        
        # Get current values (first row since data is already newest first)
        current = df.iloc[0]
        
        return {
            'sma50': float(current['sma50']) if not pd.isna(current['sma50']) else 0.0,
            'sma200': float(current['sma200']) if not pd.isna(current['sma200']) else 0.0,
            'price_above_sma50': float(current['close']) > float(current['sma50']) if not pd.isna(current['sma50']) else False,
            'price_above_sma200': float(current['close']) > float(current['sma200']) if not pd.isna(current['sma200']) else False,
            'sma50_above_sma200': float(current['sma50']) > float(current['sma200']) if not pd.isna(current['sma50']) and not pd.isna(current['sma200']) else False
        }

    def calculate_rsi(self, candles, period=14):
        """Calculate RSI using ta library's RSIIndicator."""
        try:
            # Sort candles by timestamp in ascending order (oldest first)
            sorted_candles = sorted(candles, key=lambda x: x['timestamp'])
            
            # Convert to pandas DataFrame and extract close prices
            df = pd.DataFrame(sorted_candles)
            price_series = pd.Series([float(p) for p in df['close']])
            
            # Calculate RSI using ta library
            rsi_indicator = RSIIndicator(
                close=price_series,
                window=period,
                fillna=True
            )
            rsi = rsi_indicator.rsi()
            
            # Debug logging
            logger.info('\nLast 10 positions of RSI calculation:')
            for i in range(min(10, len(price_series))):
                logger.info(
                    f"Position -{i}: "
                    f"Price: {price_series.iloc[-(i+1)]:.2f}, "
                    f"RSI: {rsi.iloc[-(i+1)]:.2f}"
                )
            
            logger.info(f"Number of prices: {len(price_series)}, Number of RSI values: {len(rsi)}")
            
            return rsi
            
        except Exception as e:
            logger.error(f"Error calculating RSI: {str(e)}")
            return pd.Series()  # Return empty series on error

    def calculate_rsi_divergences(self, candles: List[Dict[str, Any]], lookback_left: int = 5, lookback_right: int = 5) -> Dict[str, bool]:
        try:
            signals = {
                'bullish_divergence': False,
                'hidden_bullish_divergence': False,
                'bearish_divergence': False,
                'hidden_bearish_divergence': False
            }

            if len(candles) < 20:  # Minimum candles needed for reliable RSI and divergence calculation
                logger.info("Not enough candles for RSI divergence calculation")
                return signals

            # Convert candles to DataFrame
            df = pd.DataFrame(sorted(candles, key=lambda x: x['timestamp'], reverse=True))
            df = df.reset_index(drop=True)  # Reset index to avoid comparison issues
            
            # Calculate RSI
            rsi_series = self.calculate_rsi(candles)
            rsi_series = rsi_series.reset_index(drop=True)  # Reset index to avoid comparison issues

            # Debug logging
            logger.info(f"RSI Divergence calculation - Number of candles: {len(candles)}")
            logger.info(f"RSI Series length: {len(rsi_series)}")

            # Pivot points
            pl_found = rsi_series.rolling(window=lookback_left + lookback_right + 1).min().shift(-lookback_right).notna()
            ph_found = rsi_series.rolling(window=lookback_left + lookback_right + 1).max().shift(-lookback_right).notna()

            if len(pl_found) > 0 and len(ph_found) > 0:
                # Regular Bullish Divergence
                rsi_hl = (rsi_series.shift(lookback_right) > rsi_series.shift(lookback_right + 1)) & pl_found
                price_ll = (df['low'].shift(lookback_right) < df['low'].shift(lookback_right + 1))
                bull_cond = price_ll & rsi_hl & pl_found

                signals['bullish_divergence'] = bull_cond.any()

                # Hidden Bullish Divergence
                rsi_ll = (rsi_series.shift(lookback_right) < rsi_series.shift(lookback_right + 1)) & pl_found
                price_hl = (df['low'].shift(lookback_right) > df['low'].shift(lookback_right + 1))
                hidden_bull_cond = price_hl & rsi_ll & pl_found

                signals['hidden_bullish_divergence'] = hidden_bull_cond.any()

                # Regular Bearish Divergence
                rsi_lh = (rsi_series.shift(lookback_right) < rsi_series.shift(lookback_right + 1)) & ph_found
                price_hh = (df['high'].shift(lookback_right) > df['high'].shift(lookback_right + 1))
                bear_cond = price_hh & rsi_lh & ph_found

                signals['bearish_divergence'] = bear_cond.any()

                # Hidden Bearish Divergence
                rsi_hh = (rsi_series.shift(lookback_right) > rsi_series.shift(lookback_right + 1)) & ph_found
                price_lh = (df['high'].shift(lookback_right) < df['high'].shift(lookback_right + 1))
                hidden_bear_cond = price_lh & rsi_hh & ph_found

                signals['hidden_bearish_divergence'] = hidden_bear_cond.any()

            return signals

        except Exception as e:
            logger.error(f"Error calculating RSI Divergences: {str(e)}")
            return {
                'bullish_divergence': False,
                'hidden_bullish_divergence': False,
                'bearish_divergence': False,
                'hidden_bearish_divergence': False
            }

# Create singleton instance
indicators = Indicators()